File: 3955_valid_binary_strings_with_cost_limit/solution.py
# LeetCode 3955 - Valid Binary Strings With Cost Limit
# https://leetcode.com/problems/valid-binary-strings-with-cost-limit/

from typing import List


class Solution:
    def generateValidStrings(self, n: int, k: int) -> List[str]:
        ans: List[str] = []
        path: List[str] = []
        self.dfs(0, 0, n, k, path, ans)
        return ans

    def dfs(self, i: int, tot: int, n: int, k: int, path: List[str], ans: List[str]) -> None:
        if i >= n:
            ans.append("".join(path))
            return
        path.append("0")
        self.dfs(i + 1, tot, n, k, path, ans)
        path.pop()
        if (len(path) == 0 or path[-1] == "0") and tot + i <= k:
            path.append("1")
            self.dfs(i + 1, tot + i, n, k, path, ans)
            path.pop()

File: 3955_valid_binary_strings_with_cost_limit/solution.rs
// LeetCode 3955 - Valid Binary Strings With Cost Limit
// https://leetcode.com/problems/valid-binary-strings-with-cost-limit/

impl Solution {
    pub fn generate_valid_strings(n: i32, k: i32) -> Vec<String> {
        let mut ans = Vec::new();
        let mut path = String::new();
        fn dfs(i: i32, tot: i32, n: i32, k: i32, path: &mut String, ans: &mut Vec<String>) {
            if i >= n {
                ans.push(path.clone());
                return;
            }
            path.push('0');
            dfs(i + 1, tot, n, k, path, ans);
            path.pop();
            if (path.is_empty() || path.ends_with('0')) && tot + i <= k {
                path.push('1');
                dfs(i + 1, tot + i, n, k, path, ans);
                path.pop();
            }
        }
        dfs(0, 0, n, k, &mut path, &mut ans);
        ans
    }
}
